### PR TITLE
feat(ser): Trait Abstractions

### LINKED ISSUE #20: feat(ser): Trait Abstractions
### Description

Set up trait abstractions for external data sources required as input to serialize + compress data.

- Rollup Node
  - [`sync_status() -> Result<SyncStatus>`](https://github.com/refcell/cabrinha/blob/main/crates/ser/src/traits/rollup.rs#L8-L13)
- L1 Provider
  - [`header_by_number(&self, number: u64) -> Result<Header>`](https://github.com/refcell/cabrinha/blob/main/crates/ser/src/traits/providers.rs#L14C14-L14C69)
  - [`nonce_at(&self, account: Address, block_number: u64) -> Result<u64>`](https://github.com/refcell/cabrinha/blob/main/crates/ser/src/traits/providers.rs#L17C14-L17C81)
- L2 Provider
  - [`block_ref_by_number(&self, number: u64) -> Result<L2BlockRef>`](https://github.com/refcell/cabrinha/blob/main/crates/ser/src/traits/providers.rs#L24)
  - [`block_by_number(&self, number: u64) -> Result<Block>`](https://github.com/refcell/cabrinha/blob/main/crates/ser/src/traits/providers.rs#L27C14-L27C66)

### PR DESCRIPTION
**Pull request by refcell**
### Description

> [!WARNING]
>
> Follows #22 - don't merge until #22 is in trunk

Adds trait abstractions and associated types to the `ser` crate.

### Metadata

Closes #20

### CODE REVIEW

File: crates/ser/src/types/sync.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    pub parent_hash: B256,
+    /// The timestamp.
+    pub timestamp: u64,
+}
+
+/// The [`SyncStatus`][ss] of an Optimism Rollup Node.
+///
+/// [ss]: https://github.com/ethereum-optimism/optimism/blob/develop/op-service/eth/sync_status.go#L5
+#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct SyncStatus {
```
Comment (mattsse) [inline crates/ser/src/types/sync.rs:49]: before we introduce these here,

could we instead move all of these 

https://github.com/paradigmxyz/reth/blob/c788b6a585bbbd7ed92aabc42cb612a6a804daff/crates/optimism/rpc/src/api.rs#L33-L33

to op-alloy instead and import?

but could also do this in a followup

File: crates/ser/src/types/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+impl From<&BatchTransaction> for Bytes {
+    fn from(tx: &BatchTransaction) -> Self {
+        let mut buf: Vec<u8> = Vec::new();
+        for frame in tx.frames.iter() {
+            buf.append(&mut frame.encode());
+        }
+        buf.into()
+    }
```
Comment (mattsse) [inline crates/ser/src/types/transaction.rs:36]: I'd prefer a function for this

**Review by mattsse — APPROVED** [review]
+ suggestions

File: crates/ser/src/types/sync.rs (hunk 1 above)
Comment (refcell) [reply crates/ser/src/types/sync.rs:49]: Looks like someone saw this and already put up a pr :) https://github.com/alloy-rs/op-alloy/pull/35

Comment (refcell) [reply crates/ser/src/types/sync.rs:49]: Happy to wait until that's in and re-use those types.

**Pull request closed without merging**